text/1008-tls-and-alpn.rst

```diff
@@ -0,0 +1,458 @@
+                    certificate as well as any number of CA certificates
+                    needed to establish the certificate’s authenticity.
+                    If not present, a self-signed certificate will be
+                    generated and used instead (for development only!).
+    --tls-keyfile   Path to a file containing the private key. If not
+                    present, the private key will be taken from
+                    --tls-certfile as well.
+
+The Python builtin TLS support will be used to handle the certificates
+and ALPN, and the TLS transport implementation in uvloop is used for the
+network. The ``ssl.SSLContext`` [4]_ will be initialized with the
+default ``protocol=ssl.PROTOCOL_TLS``, leaving the control of accepted
+TLS protocol versions to ``SSLContext.minimum_version`` and
+``SSLContext.maximum_version``, which in turn are managed by the
+corresponding EdgeDB configs mentioned in previous chapter, together
+with the other minor tunings for ``ssl.SSLContext``.
+
+``--tls-certfile``, ``--tls-keyfile`` are directly the parameters of
+``ssl.SSLContext.load_cert_chain()``, while the EdgeDB server would
+accept a password for the private key as an environment variable
+``EDGEDB_TLS_PRIVATE_KEY_PASSWORD``. However, the ``password`` argument
+of ``load_cert_chain()`` must always be set to a Python function to
+avoid triggering OpenSSL to prompt for password. If the env var is not
+set, simply return ``b""`` in the function - it will not be invoked if
+the private key is not protected by a password.
+
+When ``--tls-certfile`` is not present and the server is in ``devmode``
+or ``testmode``, the server will use the CLI to generate a self-signed
+certificate and use it to run the TLS server for development and
+testing. The generated certificate (with its private key) should be
+written to a ``credentials.json`` file just like the regular CLI
+behavior, together with the connection parameters. For example, the
+server would run this "hidden" CLI command on start::
+
+    edgedb _dev --port 5656 --user edgedb \
+        --password login-password-in-clear-text --database edgedb
+
+And it overwrites ``~/.edgedb/credentials/local-dev.json`` with::
+
+    {
+        "port": 5656,
+        "user": "edgedb",
+        "password": "login-password-in-clear-text",
+        "database": "edgedb",
+        "tls_certdata": "-----BEGIN RSA PRIVATE KEY-----\nMIIEowIBAA..."
+    }
+
+And echo back the generated certificate in standard output so that the
+server could simply read and use. So that the EdgeDB developer could
+always use ``edgedb -Ilocal-dev`` to access the dev server. The test
+suite could also take advantage from this mimic of real-life EdgeDB
+scenario to cover some real cases.
+
+Another server-side topic that was discussed in this RFC is the UNIX
+domain socket. It is proposed that the non-admin UNIX socket support
+should be removed, while the admin UNIX socket remains in clear-text
```

Comment:
Does this mean that we would undeprecate the currently deprecated admin unix socket features?

Reply:
I don't think so. It's still an appropriate way to use for auxiliary needs like resetting password, but I don't think we need it for anything else.

E.g. I don't think this is a good way for batch processing of data as disk usage time will probably dominate on the encryption time anyway.

Reply:
Yes, we should clarify that the support of admin UNIX socket will be left only in the CLI. All bindings will not be able to connect to UNIX sockets at all.